usr/src/lib/smbclnt/libfknsmb/common/fake_stream.c

```diff
@@ -0,0 +1,1834 @@
+	dbp->db_struioun.cksum.flags = 0;
+	return (0);
+}
+
+static int
+bcache_dblk_constructor(void *buf, void *cdrarg, int kmflags)
+{
+	dblk_t *dbp = buf;
+	bcache_t *bcp = cdrarg;
+
+	if ((dbp->db_mblk = kmem_cache_alloc(mblk_cache, kmflags)) == NULL)
+		return (-1);
+
+	dbp->db_base = kmem_cache_alloc(bcp->buffer_cache, kmflags);
+	if (dbp->db_base == NULL) {
+		kmem_cache_free(mblk_cache, dbp->db_mblk);
+		return (-1);
+	}
+
+	dbp->db_mblk->b_datap = dbp;
+	dbp->db_cache = (void *)bcp;
+	dbp->db_lim = dbp->db_base + bcp->size;
+	dbp->db_free = dbp->db_lastfree = bcache_dblk_lastfree;
+	dbp->db_frtnp = NULL;
+	dbp->db_fthdr = NULL;
+	dbp->db_credp = NULL;
+	dbp->db_cpid = -1;
+	dbp->db_struioflag = 0;
+	dbp->db_struioun.cksum.flags = 0;
+	return (0);
+}
+
+/*ARGSUSED*/
+static void
+dblk_destructor(void *buf, void *cdrarg)
+{
+	dblk_t *dbp = buf;
+	ssize_t msg_size = (ssize_t)cdrarg;
+
+	ASSERT(dbp->db_mblk->b_datap == dbp);
+	ASSERT(msg_size != 0);
+	ASSERT(dbp->db_struioflag == 0);
+	ASSERT(dbp->db_struioun.cksum.flags == 0);
+
+	if ((msg_size & PAGEOFFSET) == 0) {
+		kmem_free(dbp->db_base, msg_size);
+	}
+
+	kmem_cache_free(mblk_cache, dbp->db_mblk);
+}
+
+static void
+bcache_dblk_destructor(void *buf, void *cdrarg)
+{
+	dblk_t *dbp = buf;
+	bcache_t *bcp = cdrarg;
+
+	kmem_cache_free(bcp->buffer_cache, dbp->db_base);
+
+	ASSERT(dbp->db_mblk->b_datap == dbp);
+	ASSERT(dbp->db_struioflag == 0);
+	ASSERT(dbp->db_struioun.cksum.flags == 0);
+
+	kmem_cache_free(mblk_cache, dbp->db_mblk);
+}
+
+#ifdef	_KERNEL
+
+/* ARGSUSED */
+static int
+ftblk_constructor(void *buf, void *cdrarg, int kmflags)
+{
+	ftblk_t *fbp = buf;
+	int i;
+
+	bzero(fbp, sizeof (ftblk_t));
+	if (str_ftstack != 0) {
+		for (i = 0; i < FTBLK_EVNTS; i++)
+			fbp->ev[i].stk = kmem_alloc(sizeof (ftstk_t), kmflags);
+	}
+
+	return (0);
+}
+
+/* ARGSUSED */
+static void
+ftblk_destructor(void *buf, void *cdrarg)
+{
+	ftblk_t *fbp = buf;
+	int i;
+
+	if (str_ftstack != 0) {
+		for (i = 0; i < FTBLK_EVNTS; i++) {
+			if (fbp->ev[i].stk != NULL) {
+				kmem_free(fbp->ev[i].stk, sizeof (ftstk_t));
+				fbp->ev[i].stk = NULL;
+			}
+		}
+	}
+}
+
+static int
+fthdr_constructor(void *buf, void *cdrarg, int kmflags)
+{
+	fthdr_t *fhp = buf;
+
+	return (ftblk_constructor(&fhp->first, cdrarg, kmflags));
+}
+
+static void
+fthdr_destructor(void *buf, void *cdrarg)
+{
+	fthdr_t *fhp = buf;
+
+	ftblk_destructor(&fhp->first, cdrarg);
+}
+
+#endif	/* _KERNEL */
+
+/* Needed in the ASSERT below */
+#ifdef	DEBUG
+#ifdef	_KERNEL
+#define	KMEM_SLAB_T_SZ	sizeof (kmem_slab_t)
+#else	/* _KERNEL */
+#define	KMEM_SLAB_T_SZ	64	/* fakekernel */
+#endif	/* _KERNEL */
+#endif	/* DEBUG */
+
+void
+streams_msg_init(void)
+{
+	char name[40];
+	size_t size;
+	size_t lastsize = DBLK_MIN_SIZE;
+	size_t *sizep;
+	struct kmem_cache *cp;
+	size_t tot_size;
+	int offset;
+
+	mblk_cache = kmem_cache_create("streams_mblk", sizeof (mblk_t), 32,
+	    NULL, NULL, NULL, NULL, NULL, mblk_kmem_flags);
```

Comment:
The more I think about this, the more I wonder why we need to bother with the whole separate caches for kernel space mblks.  Is there value in keeping this complexity for your emulation layer?

Reply:
Was just doing what got me there in the shortest amount of time. :)